runtime/common/src/migrations/hold_reason.rs

```diff
@@ -0,0 +1,151 @@
+// (at your option) any later version (see http://www.gnu.org/licenses).
+// Centrifuge is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+use cfg_primitives::{AccountId, Balance};
+use frame_support::{
+	pallet_prelude::ValueQuery,
+	storage_alias,
+	traits::{ConstU32, Get, Len, OnRuntimeUpgrade, VariantCountOf},
+	Blake2_128Concat, BoundedVec, Parameter,
+};
+use pallet_balances::IdAmount;
+use pallet_order_book::weights::Weight;
+use pallet_transfer_allowlist::HoldReason;
+#[cfg(feature = "try-runtime")]
+use parity_scale_codec::{Decode, Encode};
+use parity_scale_codec::{FullCodec, FullEncode};
+#[cfg(feature = "try-runtime")]
+use sp_runtime::Saturating;
+#[cfg(feature = "try-runtime")]
+use sp_runtime::TryRuntimeError;
+use sp_runtime::{traits::Member, SaturatedConversion};
+use sp_std::{vec, vec::Vec};
+
+const LOG_PREFIX: &str = "MigrateTransferAllowList HoldReason:";
+
+pub struct MigrateTransferAllowListHolds<T, RuntimeHoldReason>(
+	sp_std::marker::PhantomData<(T, RuntimeHoldReason)>,
+);
+
+type OldHolds = BoundedVec<IdAmount<(), Balance>, ConstU32<10>>;
+type NewHolds<RuntimeHoldReason> =
+	BoundedVec<IdAmount<RuntimeHoldReason, Balance>, VariantCountOf<RuntimeHoldReason>>;
+
+#[storage_alias]
+pub type Holds<T: pallet_balances::Config> =
+	StorageMap<pallet_balances::Pallet<T>, Blake2_128Concat, AccountId, OldHolds, ValueQuery>;
+
+impl<T, RuntimeHoldReason> OnRuntimeUpgrade for MigrateTransferAllowListHolds<T, RuntimeHoldReason>
+where
+	T: pallet_balances::Config<Balance = Balance, RuntimeHoldReason = RuntimeHoldReason>
+		+ pallet_transfer_allowlist::Config
+		+ frame_system::Config<AccountId = AccountId>,
+	<T as pallet_balances::Config>::RuntimeHoldReason: From<pallet_transfer_allowlist::HoldReason>,
+	RuntimeHoldReason: frame_support::traits::VariantCount
+		+ FullCodec
+		+ FullEncode
+		+ Parameter
+		+ Member
+		+ sp_std::fmt::Debug,
+{
+	fn on_runtime_upgrade() -> Weight {
+		let transfer_allowlist_accounts: Vec<AccountId> =
+			pallet_transfer_allowlist::AccountCurrencyTransferAllowance::<T>::iter_keys()
+				.map(|(a, _, _)| a)
+				.collect();
+		let mut weight =
+			T::DbWeight::get().reads(transfer_allowlist_accounts.len().saturated_into());
+
+		pallet_balances::Holds::<T>::translate::<OldHolds, _>(|who, holds| {
+			if Self::account_can_be_migrated(&who, &transfer_allowlist_accounts) {
+				log::info!("{LOG_PREFIX} Migrating hold for account {who:?}");
+				let id_amount = IdAmount::<RuntimeHoldReason, Balance> {
+					id: HoldReason::TransferAllowance.into(),
+					// Non-emptiness ensured above
+					amount: holds[0].amount,
+				};
+				weight.saturating_accrue(T::DbWeight::get().reads_writes(1, 1));
+
+				Some(NewHolds::truncate_from(vec![id_amount]))
+			} else {
+				None
+			}
```

Comment:
I hope that one is not too heavy. How many entries do we have again?

Reply:
Just two on production luckily. They can only emerge from transfer allowlist for native currency. For more info, see here https://github.com/centrifuge/centrifuge-chain/pull/1849#discussion_r1622051322